src/server/api/routes/users/v2/event.js

```diff
@@ -1,20 +1,41 @@
 const { Router } = require('express');
 const {
   walletUserCollection,
-  FieldValue,
+  Timestamp,
 } = require('../../../../modules/firebase');
 const { authenticateV2Login } = require('../../../middleware/auth');
 const { setPrivateCacheHeader } = require('../../../middleware/cache');
 const { publisher, PUBSUB_TOPIC_MISC } = require('../../../../modules/pubsub');
 
 const router = Router();
 
+function isValidTs(ts, lastSeenTs) {
+  // eslint-disable-next-line no-restricted-globals
+  if (isNaN(ts) || ts % 1000 !== 0) {
+    return false;
+  }
+  if (ts < lastSeenTs) {
+    return false;
+  }
+  if (ts > Date.now() + 5 * 60 * 1000) {
+    return false;
+  }
+  return true;
+}
+
 router.post('/event/seen', authenticateV2Login, async (req, res, next) => {
+  const { ts } = req.body;
   try {
     setPrivateCacheHeader(res);
     const { user } = req.session;
+    const userDoc = await walletUserCollection.doc(user).get();
+    const { eventLastSeenTs } = userDoc.data();
+    if (!isValidTs(ts, eventLastSeenTs)) {
+      res.status(400).send('INVALID_TIMESTAMP_PROVIDED');
+      return;
+    }
     await walletUserCollection.doc(user).update({
-      eventLastSeenTs: FieldValue.serverTimestamp(),
+      eventLastSeenTs: Timestamp.fromMillis(Math.floor(ts / 1000) * 1000),
     });
```

Comment:
Note that this `get()` then` update()` check normally requires a transaction, but it kind of doesn't matter here